Python/Product/Analysis/Interpreter/SentinelModule.cs

```diff
@@ -36,12 +39,12 @@ sealed class SentinelModule : IPythonModule {
             }
         }
 
-        public async Task<IPythonModule> WaitForImportAsync(CancellationToken cancellationToken) {
+        public async Task<IPythonModule> WaitForImportAsync(IPythonInterpreter interpreter, CancellationToken cancellationToken) {
             var mod = _realModule;
             if (mod != null) {
                 return mod;
             }
-            if (_thread == Thread.CurrentThread) {
+            if (_thread == Thread.CurrentThread || _interpreter == interpreter) {
```

Comment:
I don't understand how it works:
- Thread checking doesn't work in async environment, it requires `AsyncLocal`.
- `IPythonInterpreter` instances are shared between requests, how do they help against reentrancy?

Reply:
Yeah, now I think about it, I'm not sure `IPythonInterpreter` is the right check here. I'd need to think through what is, it certainly isn't obvious. (The aim is to ensure parallel tasks will wait for some time for an import to complete, but not if it is recursing and will never make progress. `AsyncLocal` is probably correct, though I haven't convinced myself of that yet.)

Reply:
I left the current thread check for legacy reasons. If it is safe to remove, I can nix it. The intent here is to determine that code is reentering on import (i.e. circular import). So sentinel remember which interpreter it relates to.

Reply:
`_thread == Thread.CurrentThread` may become false positive, i.e. sentinel was created on thread A, then some importing code was started on thread B but async was continued on thread A.

In `AsyncLocal` we can store the modules that requested this import. But I don't fully understand how the sentinel works in case of reentrancy. If `WaitForImportAsync` has returned the sentinel itself, what happens next?

Reply:
Yeah, I know about the thread issue, this is what we are trying to solve. `WaitForImportAsync` is called if some code requests the import and module table contains sentinel (i.e. module is being imported).

Reply:
`SynchronizationContext.Current` might be the right comparison. We should always reach this point on the "main" analysis thread, since imports should be synchronous if they are already cached, and we're aren't doing locking here as much as we are detecting same-thread vs. other-thread.

That said, I can see it being a maintenance issue in the future. What if we store the current synchronization context on this object, but check it from the caller? That way "everyone" who comes into this function will get a valid result (or wait), and we can handle the `ImportModule` recursion in `ImportModule` (or whichever helper function we're calling this from).

Reply:
`_interpreter.ImportModule` which calls `TryImportModuleAsync` on sync code path is called from `Task.Run(() => _interpreter.ImportModule(name))` - there is no sync context in it.

Reply:
Hmm... you're right. But that also means that AsyncLocal wouldn't work either, right?

Reply:
No, `AsyncLocal` has nothing to do with `SynchronizationContext`, it works through `ExecutionContext`.

Reply:
`AsyncLocal` does not work. Actually, the interpreter instance appears to be a good choice. Import context (`TryImportModuleContext`) consists of interpreter and some of its members (such as module table). So the importing task is started in `ImportModule` for a given name and context which is, in fact, the interpreter. 

Similar effect can be achieved by tracking which import tasks already started in the `ImportModule` and then avoiding starting new ones for the same name/context pair. But it looks like same thing as checking the interpreter instance since context is built around it.

So, if we are importing into module table that is per interpreter instance, then checking said instance should be OK. No?